MDEV-34919 - Aria crashes with high-level (vector) indexes

svoj opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-34919*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description
Fixes aria crash with high-level indexes.

## Release Notes
None.

## How can this PR be tested?
mtr vector

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and codying style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [ ] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [ ] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

dr-m committed:
10.11: MDEV-34062: Implement innodb_log_file_mmap on 64-bit systems

When using the default innodb_log_buffer_size=2m, mariadb-backup --backup
would spend a lot of time re-reading and re-parsing the log. For reads,
it would be beneficial to memory-map the entire ib_logfile0 to the
address space (typically 48 bits or 256 TiB) and read it from there,
both during --backup and --prepare.

We will introduce the Boolean read-only parameter innodb_log_file_mmap
that will be OFF by default on most platforms, to avoid aggressive
read-ahead of the entire ib_logfile0 in when only a tiny portion would be
accessed. On Linux and FreeBSD the default is innodb_log_file_mmap=ON,
because those platforms define a specific mmap(2) option for enabling
such read-ahead and therefore it can be assumed that the default would
be on-demand paging. This parameter will only have impact on the initial
InnoDB startup and recovery. Any writes to the log will use regular I/O,
except when the ib_logfile0 is stored in a specially configured file system
that is backed by persistent memory (Linux "mount -o dax").

We also experimented with allowing writes of the ib_logfile0 via a
memory mapping and decided against it. A fundamental problem would be
unnecessary read-before-write in case of a major page fault, that is,
when a new, not yet cached, virtual memory page in the circular
ib_logfile0 is being written to. There appears to be no way to tell
the operating system that we do not care about the previous contents of
the page, or that the page fault handler should just zero it out.

Many references to HAVE_PMEM have been replaced with references to
HAVE_INNODB_MMAP.

The predicate log_sys.is_pmem() has been replaced with
log_sys.is_mmap() && !log_sys.is_opened().

Memory-mapped regular files differ from MAP_SYNC (PMEM) mappings in the
way that an open file handle to ib_logfile0 will be retained. In both
code paths, log_sys.is_mmap() will hold. Holding a file handle open will
allow log_t::clear_mmap() to disable the interface with fewer operations.

It should be noted that ever since
commit 685d958e38b825ad9829be311f26729cccf37c46 (MDEV-14425)
most 64-bit Linux platforms on our CI platforms
(s390x a.k.a. IBM System Z being a notable exception) read and write
/dev/shm/*/ib_logfile0 via a memory mapping, pretending that it is
persistent memory (mount -o dax). So, the memory mapping based log
parsing that this change is enabling by default on Linux and FreeBSD
has already been extensively tested on Linux.

::log_mmap(): If a log cannot be opened as PMEM and the desired access
is read-only, try to open a read-only memory mapping.

xtrabackup_copy_mmap_snippet(), xtrabackup_copy_mmap_logfile():
Copy the InnoDB log in mariadb-backup --backup from a memory
mapped file.

dr-m committed:
10.11: MDEV-34907 Bogus assertion failure and busy work while parsing FILE_ records

A server that was running with innodb_log_file_size=96M and
innodb_buffer_pool_size=6M had inserted some data into a table
that was subsequently dropped. When the server was killed and
restarted, an assertion failed in recv_sys_t::parse() while
a FSP_SIZE change was unnecessarily being processed during
the skip_the_rest: loop in recv_scan_log().

The ib_logfile0 contents was as follows:

1. The checkpoint start LSN points to the start of some mini-transaction.
2. There may be log records for modifying files for which a FILE_MODIFY
had been written before the checkpoint. These records were "purged"
by advancing the checkpoint.
3. At some point during the initial parsing with store=true the space
reserved for recv_sys.pages will run out and recv_scan_log() would switch
to the skip_the_rest: mode.
4. We encounter a log record for extending a tablespace that will be
deleted a bit later. This would trip the bogus debug assertion.
5. Later on, there would be a FILE_DELETE record for this tablespace.
6. The checkpoint end LSN points to a possibly empty sequence of
FILE_MODIFY records and a FILE_CHECKPOINT record. Recovery had parsed these
records first, before rewinding to the checkpoint start LSN.
7. There could be further records following the FILE_CHECKPOINT record.
Recovery will process all records until an inconsistency is found and
it is assumed that the end of the circular ib_logfile0 was reached.

recv_sys_t::parse(): For the template instantiation with store=false,
remove a debug assertion that could fail in a multi-batch recovery,
while recv_scan_log(false) would be in the skip_the_rest: loop.
It is very well possible that we have not encountered all FILE_ records
yet, and therefore we should not complain about unknown tablespaces.

Reviewed by: Debarun Banerjee

dr-m committed:
10.11: MDEV-34450 os_file_write_func() is an overkill for ib_logfile0

log_file_t::read(), log_file_t::write(): Invoke pread() or pwrite()
directly, so that we can give more accurate diagnostics in case of
a failure, and so that we will avoid the overhead of setting up 5(!)
stack frames and related objects.

tpool::pwrite(): Add a missing const qualifier.

dr-m committed:
10.11: MDEV-34850: Busy work while parsing FILE_ records

In mariadb-backup --backup, we only have to invoke the undo_space_trunc
and log_file_op callbacks as well as validate the mini-transaction
checksums. There is absolutely no need to access recv_sys.pages or
recv_spaces, or to allocate a decrypt_buf in case of innodb_encrypt_log=ON.
This is what the new mode recv_sys_t::store::BACKUP will do.

In the skip_the_rest: loop, the main thing is to process all FILE_ records
until the end of the log is reached.  Additionally, we must process
INIT_PAGE and FREE_PAGE records in the same way as they would be
during storing == YES.

This was measured to reduce the CPU time between the messages
"InnoDB: Multi-batch recovery needed at LSN" and
"InnoDB: End of log at LSN"
by some 20%.

recv_sys_t::store: A ternary enumeration that specifies how records
should be stored: NO, BACKUP, or YES.

recv_sys_t::parse(), recv_sys_t::parse_mtr(), recv_sys_t::parse_pmem():
Replace template<bool store> with template<store storing>.

store_freed_or_init_rec(): Simplify some logic. We can look up also
the system tablespace.

Reviewed by: Debarun Banerjee

vuvova committed:
MDEV-35092 Server crash, hang or ASAN errors in mysql_create_frm_image upon using non-default table options and system variables

extend the option_list expicitly on CREATE/ALTER, not implicitly
on parsing.

vuvova committed:
MDEV-35104 Invalid (old?) table or database name upon DDL on table with vector key and unique key

InnoDB rename needs the same workaround for hlindexes
as it has for partitions

vuvova committed:
MDEV-35105 Assertion `tab->join->order' fails upon vector search with DISTINCT

don't apply distinct optimization to order by a vector index

svoj committed:
MDEV-34919 - Aria crashes with high-level (vector) indexes

Since high-level index tables do not participate in thr_multi_lock(), added
explicit call to THR_LOCK::start_trans(). This is needed mostly for Aria to
handle transaction logging.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3570) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3570) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: svoj<br/>:x: vuvova<br/>:x: dr-m<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3570) it.</sub>

vuvova committed:
MDEV-33410 VECTOR data type

vuvova committed:
MDEV-35141 Server crashes in Field_vector::report_wrong_value upon statistic collection

vuvova committed:
MDEV-35147 Inconsistent NULL handling in vector type

vuvova committed:
MDEV-35150 Column containing non-vector tables can be modified to VECTOR type without warnings

vuvova committed:
MDEV-35158 Assertion `res->length() > 0 && res->length() % 4 == 0' fails upon increasing length of vector column

vuvova committed:
Vec_ToText was underestimating max_length of the result

switch to a more predictable, shorter, and more correct output
that is, print as many significant digits as necessary.
but not more (they'd be just zeros) and not less (it'd lose precision)

vuvova committed:
cleanup: remove String::append_float

between set_real() and set_fcvt(), a third String method for
real->string conversion looks definitely redundant

vuvova committed:
MDEV-35159 Assertion `tab->join->select_limit < (~ (ha_rows) 0)' fails upon forcing vector key

init_from_binary_frm_image() wrongly assumed that
* if a table has primary key
* and it has the HA_PRIMARY_KEY_IN_READ_INDEX flag
* than ORDER BY any index automatically implies ORDER BY pk at the end,
   that is for an index (a,b,c) ORDER BY a,b,c means ORDER BY a,b,c,pk

which is wrong, it holds not for _any index_ but only for indexes
that can be used for ORDER BY.

So, don't do `field->part_of_sortkey= share->keys_in_use`
but introduce `sort_keys_in_use` and use that.

vuvova committed:
MDEV-35160 RBR does not work with vector type, ER_SLAVE_CONVERSION_FAILED

test case

vuvova committed:
add MYSQL_TYPE_VECTOR for MySQL compatibility

internal and replication

vuvova committed:
add MYSQL_TYPE_VECTOR for MySQL compatibility. protocol

vuvova committed:
MDEV-35178 Assertion failure in Field_vector::store upon INSERT IGNORE with a wrong data

vuvova committed:
fixup! MDEV-35092

vuvova committed:
MDEV-35176 ASAN errors in Field_vector::store with optimizer_trace enabled

vuvova committed:
MDEV-35177 Unexpected ER_TRUNCATED_WRONG_VALUE_FOR_FIELD, diagnostics area assertion failures upon EITS collection with vector type

don't collect histograms for vector columns

vuvova committed:
MDEV-35182 crash in online_alter_end_trans with XA over vector indexes

ONLINE ALTER didn't expect XA PREPARE to fail.
Mark rollback on failed prepare with the XA_ROLLBACK_ONLY state,
detect that in ONLINE ALTER

vuvova committed:
MDEV-35191 Assertion failure in Create_tmp_table::finalize upon DISTINCT with vector type

add MYSQL_TYPE_VECTOR to an if() over field->type() values

vuvova committed:
Revert "add MYSQL_TYPE_VECTOR for MySQL compatibility. protocol"

This reverts commit 79b7f896d13ae2d47b561aab2786472e6323641e.

vuvova committed:
Revert "add MYSQL_TYPE_VECTOR for MySQL compatibility"

This reverts commit db658f9ba720cec0ca5f9a07d66487e3eeb31a29.

vuvova committed:
MDEV-35160 RBR does not work with vector type, ER_SLAVE_CONVERSION_FAILED

vuvova committed:
MDEV-35177 Unexpected ER_TRUNCATED_WRONG_VALUE_FOR_FIELD, diagnostics area assertion failures upon EITS collection with vector type

vuvova committed:
MDEV-35204 mysqlbinlog --verbose fails on row events with vector type

test case

vuvova committed:
MDEV-35195 Assertion `tab->join->order' fails upon vector search with DISTINCT #2

mhnsw index can be used only for ORDER BY. not for GROUP BY

vuvova committed:
MDEV-35203 ASAN errors or assertion failures in row_sel_convert_mysql_key_to_innobase upon query from table with usual key on vector field

add test

vuvova committed:
MDEV-35194 non-BNL join fails on assertion

with streaming implemened mhnsw no longer needs to know
the LIMIT in advance. let's just cap it to avoid allocating
too much memory for the one step result set

vuvova committed:
MDEV-35212 Server crashes in Item_func_vec_fromtext::val_str upon query from empty table

vuvova committed:
MDEV-35210 Vector type cannot store values which VEC_FromText produces and VEC_ToText accepts

let VEC_FromText validate that the vector l2squared isn't NaN.
VEC_ToText still prints everything.

vuvova committed:
MDEV-35205 Server crash in online alter upon concurrent ALTER and DML on table with vector field

test case

vuvova committed:
MDEV-35213 Server crash or assertion failure upon query with high value of mhnsw_min_limit

mhnsw_min_limit must not be larger than candidates queue size

vuvova committed:
MDEV-35215 ASAN errors in Item_func_vec_fromtext::val_str upon VEC_FROMTEXT with an invalid argument

vuvova commented:
Looks good. I've generalized it a bit, which allowed me to remove my InnoDB hack (that was needed because of missing store_lock).

Pushed into the vector branch. Closing.

svoj commented:
@vuvova, fine with me. Just one thing re 07b74febdcd16b5bb1c72ed9015bab51493a4e08: `store_lock()` may return multiple `THR_LOCK_DATA`, but you moved it out of `if (lock_count() == 1)` guard. Do you feel like it is not going to be relevant for this code ever? May be at least `DBUG_ASSERT(lock_count() <= 1)`?

vuvova commented:
I have an assert below, and I believe effect is the same. But I'll move it up indeed, it'll be a bit more obvious
